### PR TITLE
xonfig xontribs

### PR DESCRIPTION
**Pull request by anki-code**
_Originally posted by @scopatz in https://github.com/xonsh/xonsh/issues/3864#issuecomment-710591574_:
> We really need to have `xonfig` print out what xontribs have been loaded.

### CODE REVIEW

**Review by gforsyth — CHANGES_REQUESTED** [review]
Thanks for taking this on, @anki-code ! 

This isn't working on my end -- I don't think that `xontribs` necessarily add the `_xontribs` key to `__xonsh__.ctx` on load.  I think the more robust check is to follow the logic in `xonsh.xontribs._list` which checks if a given `spec.name in sys.modules`.

Also, I think the `xontrib` section of `xonfig` should still be displayed even if no xontribs are loaded (we want to disambiguate a user report where the version of xonsh doesn't _include_ the xontrib info vs. one where the user has no xontribs loaded)

**Comment by anki-code** [comment]
@gforsyth logic changes.

Example with xontribs:

<details>

```
$ xonfig
+------------------+--------------------------------------------+
| xonsh            | 0.9.24.dev337                              |
| Git SHA          | dd87a57c                                   |
| Commit Date      | Oct 17 18:31:44 2020                       |
| Python           | 3.8.2                                      |
| PLY              | 3.11                                       |
| have readline    | True                                       |
| prompt toolkit   | 3.0.7                                      |
| shell type       | prompt_toolkit                             |
| pygments         | 2.6.1                                      |
| on posix         | True                                       |
| on linux         | True                                       |
| distro           | ubuntu                                     |
| on darwin        | False                                      |
| on windows       | False                                      |
| on cygwin        | False                                      |
| on msys2         | False                                      |
| is superuser     | False                                      |
| default encoding | utf-8                                      |
| xonsh encoding   | utf-8                                      |
| encoding errors  | surrogateescape                            |
| on jupyter       | True                                       |
| jupyter kernel   | /opt/miniconda/share/jupyter/kernels/xonsh |
| xontrib 1        | prompt_bar                                 |
| xontrib 2        | output_search                              |
| xontrib 3        | whole_word_jumping                         |
| xontrib 4        | pipeliner_parallel                         |
| xontrib 5        | pipeliner                                  |
| xontrib 6        | argcomplete                                |
| xontrib 7        | onepath                                    |
| xontrib 8        | sh                                         |
| xontrib 9        | back2dir                                   |
| xontrib 10       | autojump                                   |
| xontrib 11       | conda                                      |
+------------------+--------------------------------------------+
```
</details>

Example without xontribs:
<details>

```
$ xonfig
+------------------+--------------------------------------------+
| xonsh            | 0.9.24.dev337                              |
| Git SHA          | dd87a57c                                   |
| Commit Date      | Oct 17 18:31:44 2020                       |
| Python           | 3.8.2                                      |
| PLY              | 3.11                                       |
| have readline    | True                                       |
| prompt toolkit   | 3.0.7                                      |
| shell type       | prompt_toolkit                             |
| pygments         | 2.6.1                                      |
| on posix         | True                                       |
| on linux         | True                                       |
| distro           | ubuntu                                     |
| on darwin        | False                                      |
| on windows       | False                                      |
| on cygwin        | False                                      |
| on msys2         | False                                      |
| is superuser     | False                                      |
| default encoding | utf-8                                      |
| xonsh encoding   | utf-8                                      |
| encoding errors  | surrogateescape                            |
| on jupyter       | True                                       |
| jupyter kernel   | /opt/miniconda/share/jupyter/kernels/xonsh |
| xontrib          |                                            |
+------------------+--------------------------------------------+
```

</details>

File: xonsh/readline_shell.py [hunk 1]
```diff
@@ -29,12 +29,7 @@
     ansi_color_style,
 )
 from xonsh.prompt.base import multiline_prompt
-from xonsh.tools import (
-    print_exception,
-    to_bool,
-    columnize,
-    carriage_return,
-)
+from xonsh.tools import print_exception, to_bool, columnize, carriage_return
```
Comment (laloch) [inline xonsh/readline_shell.py:32]: Is this unrelated change intentional?

Comment (anki-code) [reply xonsh/readline_shell.py:32]: No, it's mr. Black.

File: xonsh/xonfig.py [hunk 2]
```diff
@@ -537,6 +550,9 @@ def _info(ns):
         pass
     data.extend([("on jupyter", jup_ksm is not None), ("jupyter kernel", jup_kernel)])
 
+    xontribs = [s.split(".")[1] for s in sys.modules if s.startswith("xontrib.")]
```
Comment (scopatz) [inline xonsh/xonfig.py:553]: I think there as API for getting the xontrib list. That would be good to use here

File: xonsh/xonfig.py [hunk 3]
```diff
@@ -478,12 +478,22 @@ def _xonfig_format_human(data):
...
+            for subval in val:
+                wcol2 = max(wcol2, len(str(subval)))
+        else:
+            wcol2 = max(wcol2, len(str(val)))
     hr = "+" + ("-" * (wcol1 + 2)) + "+" + ("-" * (wcol2 + 2)) + "+\n"
     row = "| {key!s:<{wcol1}} | {val!s:<{wcol2}} |\n"
     s = hr
     for key, val in data:
-        s += row.format(key=key, wcol1=wcol1, val=val, wcol2=wcol2)
+        if type(val) == list:
```
Comment (gforsyth) [inline xonsh/xonfig.py]:
```suggestion
        if isinstance(val, list):
```

File: xonsh/xonfig.py [hunk 4]
```diff
@@ -478,12 +478,22 @@ def _xonfig_format_human(data):
     wcol1 = wcol2 = 0
     for key, val in data:
         wcol1 = max(wcol1, len(key))
-        wcol2 = max(wcol2, len(str(val)))
+        if type(val) == list:
```
Comment (gforsyth) [inline xonsh/xonfig.py:481]:
```suggestion
        if isinstance(val, list):
```

Comment (anki-code) [reply xonsh/xonfig.py]: Changed. Thanks!

File: xonsh/xonfig.py (hunk 3 above)
Comment (anki-code) [reply xonsh/xonfig.py]: Changed. Thanks!

File: xonsh/xonfig.py (hunk 2 above)
Comment (anki-code) [reply xonsh/xonfig.py]: Added + test.

File: xonsh/xontribs.py [hunk 5]
```diff
@@ -132,6 +132,11 @@ def xontribs_load(names, verbose=False):
     return res
 
 
+def xontribs_loaded():
+    """Returns list of loaded xontribs."""
+    return [s.split(".")[1] for s in sys.modules if s.startswith("xontrib.")]
```
Comment (gforsyth) [inline xonsh/xontribs.py:137]: I would rather this not iterate over the entirety of `sys.modules`

maybe something like

```
meta = xontrib_metadata()
for md in meta["xontribs"]:
    [spec.name in sys.modules for spec in find_xontrib(md["name"])
```

Comment (anki-code) [reply xonsh/xontribs.py:137]: My key idea is to show all xontribs. Not only that described in xontrib file.

Comment (anki-code) [reply xonsh/xontribs.py:137]: I also tried to use `importlib` and `pkgutil` for example:
```
import pkgutil
for importer, modname, ispkg in pkgutil.iter_modules(importlib.util.find_spec("xontrib").submodule_search_locati
ons):
     print("Found submodule %s (is a package: %s)" % (modname, ispkg))
```
But it is not working for xontribs with dash in the name. I.e. `readable-traceback`.

Comment (scopatz) [reply xonsh/xontribs.py:137]: Sure, but I think maybe xontribs that want to be found should add themselves to `xontribs_metadata`.  Maybe we should add a registration function for that.

Comment (anki-code) [reply xonsh/xontribs.py:137]: > xontribs that want to be found should add themselves

This is somewhat contrary to the goal:

> we want to disambiguate a user report where the version of xonsh doesn't include the xontrib info

If user played with some xontribs from github or try to write his own xontrib and foreget about it and then he breaked the xonsh it should run `xonfig` and remeber about his experiments.

Comment (scopatz) [reply xonsh/xontribs.py:137]: Yeah, but I think we want to be fast and stable before being fully complete. This is a piece of debugging code, so it needs to work even when other parts of xonsh/python are falling down. Does that make sense?

Comment (scopatz) [reply xonsh/xontribs.py:137]: Also, I think that once modules are loaded once, they are added to the metadata, so that `xontrib list` will show them

Comment (anki-code) [reply xonsh/xontribs.py:137]: We just need a simplest way to get the list of loaded xontribs. If `importlib` can't help because of missing the xontribs with dash in the name and `xontribs` module can't help because of using only list of known xontribs, I see the one way is to using `sys.modules` or similar. If you know the more optimal way to avoid `sys.modules` please point it out.

Comment (anki-code) [reply xonsh/xontribs.py:137]: @gforsyth there is a way to read files in `xontrib` directory instead of iterating `sys.modules`:
```
[x.name.split('.')[0] for x in Path(importlib.util.find_spec("xontrib").submodule_search_locations[0]).glob('*')
]
```
What do you think?

Comment (scopatz) [reply xonsh/xontribs.py:137]: Yeah, I think this is better. If you look at what [list does below](https://github.com/xonsh/xonsh/pull/3894/files#diff-0108dd9b8af7697497623c2f97c08b67536c0d7e4d593b5ce3e9656928854663R156-R161), it basically does a 

```
spec = find_xontrib(name)
loaded = spec.name in sys.modules
```

I would actually recommend splitting out the whole first part of the `_list` function into its own top level function for getting the `data` list.  Maybe `xontrib_data()`?  This function could then be reused in `xonfig`

Comment (anki-code) [reply xonsh/xontribs.py]: Done

**Review by gforsyth — APPROVED** [review]
Thanks for all your hard work on this, @anki-code !  This looks good to me.
